litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -1902,54 +1922,267 @@ protected LayoutState calculateLayoutState(
+      synchronized (mLayoutStateFutureLock) {
+        if (localLayoutStateFuture.equals(mLayoutStateFuture) && !mLayoutStateFuture.isReleased()) {
+          // Use the latest LayoutState calculation if it's the same and its LayoutState unreleased.
+          localLayoutStateFuture = mLayoutStateFuture;
+        } else {
+          // Otherwise set our calculation as the latest one.
+          mLayoutStateFuture = localLayoutStateFuture;
+        }
+      }
     }
 
-    if (lock != null) {
-      synchronized (lock) {
-        return LayoutState.calculate(
-            contextWithStateHandler,
-            root,
-            mId,
-            widthSpec,
-            heightSpec,
-            diffingEnabled,
-            previousLayoutState,
-            mCanPrefetchDisplayLists,
-            mCanCacheDrawingDisplayLists,
-            mShouldClipChildren,
-            mPersistInternalNodeTree,
-            source,
-            extraAttribution);
-      }
-    } else {
+    localLayoutStateFuture.run();
 
-      return LayoutState.calculate(
-          contextWithStateHandler,
+    final LayoutState layoutState = localLayoutStateFuture.get();
+    if (layoutState == null) {
+      // If the LayoutState is already released, recalculate.
+      return calculateLayoutState(
+          lock,
+          context,
           root,
-          mId,
           widthSpec,
           heightSpec,
           diffingEnabled,
           previousLayoutState,
-          mCanPrefetchDisplayLists,
-          mCanCacheDrawingDisplayLists,
-          mShouldClipChildren,
-          mPersistInternalNodeTree,
+          treeProps,
           source,
           extraAttribution);
     }
+    return layoutState;
+  }
+
+  private class LayoutStateFuture extends FutureTask<LayoutState> {
+
+    private static final int THREAD_PRIORITY_INVALID = Process.THREAD_PRIORITY_LOWEST +
+        Process.THREAD_PRIORITY_LESS_FAVORABLE * 100;
+
+    private final AtomicBoolean isFirstGet = new AtomicBoolean(true);
+    private final AtomicInteger runningThreadId = new AtomicInteger(-1);
+    private final ComponentContext context;
+    private final Component root;
+    private final int widthSpec;
+    private final int heightSpec;
+    private final boolean diffingEnabled;
+    @Nullable private final LayoutState previousLayoutState;
+    @Nullable private final TreeProps treeProps;
+    @Nullable private final String extraAttribution;
+    private volatile boolean isReleased = false;
+
+    private LayoutStateFuture(@Nullable final Object lock,
+        final ComponentContext context,
+        final Component root,
+        final int widthSpec,
+        final int heightSpec,
+        final boolean diffingEnabled,
+        @Nullable final LayoutState previousLayoutState,
+        @Nullable final TreeProps treeProps,
+        @CalculateLayoutSource final int source,
+        @Nullable final String extraAttribution) {
+      super(new Callable<LayoutState>() {
```

Comment:
Nesting this in the super() call feels messy, I'd prefer we extract

Reply:
True, I reduced this to just a reference to the shared function so it's cleaner now. Didn't want to create a separate class for a 1 line callable/lambda.